<a href="https://colab.research.google.com/github/Somrat390/EDA-Practise/blob/main/Transformation_Tecnique.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd

# Performing data deduplication
It is very likely that your dataframe contains duplicate rows. Removing them is essential to enhance the quality of the dataset. This can be done with the following steps:

In [ ]:
frame3 = pd.DataFrame({'column 1': ['Looping'] * 3 + ['Functions'] * 4, 'column 2': [10, 10, 22, 23, 23, 24, 24]})
frame3

,column 1,column 2
0,Looping,10
1,Looping,10
2,Looping,22
3,Functions,23
4,Functions,23
5,Functions,24
6,Functions,24


In [ ]:
frame3.duplicated()

,0
0,False
1,True
2,False
3,False
4,True
5,False
6,True


## The rows that say True are the ones that contain duplicated data

### Now, we can drop these duplicates using the drop_duplicates() method:

In [ ]:
frame4 = frame3.drop_duplicates()
frame4

,column 1,column 2
0,Looping,10
2,Looping,22
3,Functions,23
5,Functions,24


In [ ]:
frame3['column 3'] = range(7)
frame3

,column 1,column 2,column 3
0,Looping,10,0
1,Looping,10,1
2,Looping,22,2
3,Functions,23,3
4,Functions,23,4
5,Functions,24,5
6,Functions,24,6


In [ ]:
frame5 = frame3.drop_duplicates(['column 2'])

In [ ]:
frame5

,column 1,column 2,column 3
0,Looping,10,0
2,Looping,22,2
3,Functions,23,3
5,Functions,24,5


# Replacing values
Often, it is essential to find and replace some values inside a dataframe.

In [ ]:
import numpy as np
replaceFrame = pd.DataFrame({'column 1': [200., 3000., -786., 3000., 234., 444., -786., 332., 3332. ], 'column 2': range(9)})

In [ ]:
replaceFrame

,column 1,column 2
0,200.0,0
1,3000.0,1
2,-786.0,2
3,3000.0,3
4,234.0,4
5,444.0,5
6,-786.0,6
7,332.0,7
8,3332.0,8


In [ ]:
replaceFrame.replace(to_replace=-786, value=np.nan)

,column 1,column 2
0,200.0,0
1,3000.0,1
2,NaN,2
3,3000.0,3
4,234.0,4
5,444.0,5
6,NaN,6
7,332.0,7
8,3332.0,8


## we just replaced one value with the other values. We can also replace multiple values at once.

In order to do so, we display them using a list:

In [ ]:
replaceFrame = pd.DataFrame({'column 1': [200., 3000., -786., 3000., 234., 444., -786., 332., 3332. ], 'column 2': range(9)})
replaceFrame.replace(to_replace =[-786, 0], value= [np.nan, 2])

,column 1,column 2
0,200.0,2
1,3000.0,1
2,NaN,2
3,3000.0,3
4,234.0,4
5,444.0,5
6,NaN,6
7,332.0,7
8,3332.0,8


# Handling missing data

In [ ]:
data = np.arange(15, 30).reshape(5, 3)
dfx = pd.DataFrame(data, index=['apple', 'banana', 'kiwi', 'grapes', 'mango'], columns=['store1', 'store2', 'store3'])
dfx

,store1,store2,store3
apple,15,16,17
banana,18,19,20
kiwi,21,22,23
grapes,24,25,26
mango,27,28,29


In [ ]:
dfx['store4'] = np.nan
dfx.loc['watermelon'] = np.arange(15,19)
dfx.loc['oranges'] = np.nan
dfx['store5'] = np.nan
dfx['store4']['apple'] = 20.
dfx

/tmp/ipykernel_3807/2838008115.py:5: FutureWarning: ChainedAssignmentError: behaviour will change in pandas 3.0!
You are setting values through chained assignment. Currently this works in certain cases, but when using Copy-on-Write (which will become the default behaviour in pandas 3.0) this will never work to update the original DataFrame or Series, because the intermediate object on which we are setting values will behave as a copy.
A typical example is when you are setting values in a column of a DataFrame, like:

df["col"][row_indexer] = value

Use `df.loc[row_indexer, "col"] = values` instead, to perform the assignment in a single step and ensure this keeps updating the original `df`.

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

  dfx['store4']['apple'] = 20.


,store1,store2,store3,store4,store5
apple,15.0,16.0,17.0,20.0,NaN
banana,18.0,19.0,20.0,NaN,NaN
kiwi,21.0,22.0,23.0,NaN,NaN
grapes,24.0,25.0,26.0,NaN,NaN
mango,27.0,28.0,29.0,NaN,NaN
watermelon,15.0,16.0,17.0,18.0,NaN
oranges,NaN,NaN,NaN,NaN,NaN


# Nan values in panda objects

In [ ]:
dfx.isnull()

,store1,store2,store3,store4,store5
apple,False,False,False,False,True
banana,False,False,False,True,True
kiwi,False,False,False,True,True
grapes,False,False,False,True,True
mango,False,False,False,True,True
watermelon,False,False,False,False,True
oranges,True,True,True,True,True


In [ ]:
dfx.notnull()

,store1,store2,store3,store4,store5
apple,True,True,True,True,False
banana,True,True,True,False,False
kiwi,True,True,True,False,False
grapes,True,True,True,False,False
mango,True,True,True,False,False
watermelon,True,True,True,True,False
oranges,False,False,False,False,False


In [ ]:
dfx.isnull().sum()

,0
store1,1
store2,1
store3,1
store4,5
store5,7


In [ ]:
dfx.isnull().sum().sum()

np.int64(15)

## So, instead of counting the number of missing values, we can count the number of reported values

In [ ]:
dfx.count()

,0
store1,6
store2,6
store3,6
store4,2
store5,0


## Dropping missing values
One of the ways to handle missing values is to simply remove them from our dataset

In [ ]:
dfx.store4[dfx.store4.notnull()]

,store4
apple,20.0
watermelon,18.0


## The output shows that store4 only reported two items of data. Now, we can use the dropna() method to remove the rows

In [ ]:
dfx.store4.dropna()

,store4
apple,20.0
watermelon,18.0


### Note that the dropna() method just returns a copy of the dataframe by dropping the rows with NaN. The original dataframe is not changed.

###If dropna() is applied to the entire dataframe, then it will drop all the rows from the dataframe, because there is at least one NaN value in our dataframe

In [ ]:
dfx.dropna()

,store1,store2,store3,store4,store5


In [ ]:
dfx

,store1,store2,store3,store4,store5
apple,15.0,16.0,17.0,20.0,NaN
banana,18.0,19.0,20.0,NaN,NaN
kiwi,21.0,22.0,23.0,NaN,NaN
grapes,24.0,25.0,26.0,NaN,NaN
mango,27.0,28.0,29.0,NaN,NaN
watermelon,15.0,16.0,17.0,18.0,NaN
oranges,NaN,NaN,NaN,NaN,NaN


###Dropping by rows
We can also drop rows that have NaN values. To do so, we can use the how=all argument to drop only those rows entire values are entirely NaN

In [ ]:
dfx.dropna(how='all')

,store1,store2,store3,store4,store5
apple,15.0,16.0,17.0,20.0,NaN
banana,18.0,19.0,20.0,NaN,NaN
kiwi,21.0,22.0,23.0,NaN,NaN
grapes,24.0,25.0,26.0,NaN,NaN
mango,27.0,28.0,29.0,NaN,NaN
watermelon,15.0,16.0,17.0,18.0,NaN


###Dropping by columns
Furthermore, we can also pass axis=1 to indicate a check for NaN by columns.

In [ ]:
dfx.dropna(how='all', axis=1)


,store1,store2,store3,store4
apple,15.0,16.0,17.0,20.0
banana,18.0,19.0,20.0,NaN
kiwi,21.0,22.0,23.0,NaN
grapes,24.0,25.0,26.0,NaN
mango,27.0,28.0,29.0,NaN
watermelon,15.0,16.0,17.0,18.0
oranges,NaN,NaN,NaN,NaN


### Furthermore, we can also pass another argument, thresh, to specify a minimum number of NaNs that must exist before the column should be dropped:

In [ ]:
dfx.dropna(thresh=5, axis=1)

,store1,store2,store3
apple,15.0,16.0,17.0
banana,18.0,19.0,20.0
kiwi,21.0,22.0,23.0
grapes,24.0,25.0,26.0
mango,27.0,28.0,29.0
watermelon,15.0,16.0,17.0
oranges,NaN,NaN,NaN


# Mathematical operations with NaN
The pandas and numpy libraries handle NaN values differently for mathematical operations

In [ ]:
ar1 = np.array([100, 200, np.nan, 300])
ser1 = pd.Series(ar1)

ar1.mean(), ser1.mean()

(np.float64(nan), np.float64(200.0))

* When a NumPy function encounters NaN values, it returns NaN.
* Pandas, on the other hand, ignores the NaN values and moves ahead with processing. When performing the sum operation, NaN is treated as 0. If all the values are NaN, the result is also NaN.

In [ ]:
ser2 = dfx.store4
ser2.sum()

np.float64(38.0)

In [ ]:
ser2.mean()

np.float64(19.0)

In [ ]:
ser2.cumsum()  ##a sequence of partial sums of a given sequence, where each element is the sum of all previous elements plus the current one

,store4
apple,20.0
banana,NaN
kiwi,NaN
grapes,NaN
mango,NaN
watermelon,38.0
oranges,NaN


# Filling missing values
We can use the fillna() method to replace NaN values with any particular values.

In [ ]:
filledDf = dfx.fillna(0)
filledDf

,store1,store2,store3,store4,store5
apple,15.0,16.0,17.0,20.0,0.0
banana,18.0,19.0,20.0,0.0,0.0
kiwi,21.0,22.0,23.0,0.0,0.0
grapes,24.0,25.0,26.0,0.0,0.0
mango,27.0,28.0,29.0,0.0,0.0
watermelon,15.0,16.0,17.0,18.0,0.0
oranges,0.0,0.0,0.0,0.0,0.0


In [ ]:
dfx.mean()

,0
store1,20.0
store2,21.0
store3,22.0
store4,19.0
store5,NaN


Replacing the values with 0 will affect several statistics including mean, sum, and median

In [ ]:
filledDf.mean()

,0
store1,17.142857
store2,18.000000
store3,18.857143
store4,5.428571
store5,0.000000


## Note that there are slightly different values. Hence, filling with 0 might not be the optimal solution

# Backward and forward filling
NaN values can be filled based on the last known values.

# We want to fill store4 using the forward-filling technique

In [ ]:
dfx.store4.fillna(method='ffill')

/tmp/ipykernel_3807/4057730406.py:1: FutureWarning: Series.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  dfx.store4.fillna(method='ffill')


,store4
apple,20.0
banana,20.0
kiwi,20.0
grapes,20.0
mango,20.0
watermelon,18.0
oranges,18.0


# Here, from the forward-filling technique, the last known value is 20 and hence the rest of the NaN values are replaced by it.

The direction of the fill can be changed by changing method='bfill'

In [ ]:
dfx.store4.fillna(method='bfill')

/tmp/ipykernel_3807/1219575873.py:1: FutureWarning: Series.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  dfx.store4.fillna(method='bfill')


,store4
apple,20.0
banana,18.0
kiwi,18.0
grapes,18.0
mango,18.0
watermelon,18.0
oranges,NaN


## Interpolating missing values
The pandas library provides the interpolate() function both for the series and the dataframe. By default, it performs a linear interpolation of our missing values.

In [ ]:
ser3 = pd.Series([100, np.nan, np.nan, np.nan, 292])
ser3.interpolate()

,0
0,100.0
1,148.0
2,196.0
3,244.0
4,292.0


### Are you wondering how these values are calculated? Well, it is done by taking the first value before and after any sequence of the NaN values. In the preceding series, ser3, the first and the last values are 100 and 292 respectively. Hence, it calculates the next value as (292-100)/(5-1) = 48. So, the next value after 100 is 100 + 48 = 148.

In [ ]:
data = np.arange(15).reshape((3,5))
indexers = ['Rainfall', 'Humidity', 'Wind']
dframe1 = pd.DataFrame(data, index=indexers, columns=['Bergen', 'Oslo', 'Trondheim', 'Stavanger', 'Kristiansand'])
dframe1

,Bergen,Oslo,Trondheim,Stavanger,Kristiansand
Rainfall,0,1,2,3,4
Humidity,5,6,7,8,9
Wind,10,11,12,13,14


In [ ]:
dframe1.index = dframe1.index.map(str.upper)
dframe1

,Bergen,Oslo,Trondheim,Stavanger,Kristiansand
RAINFALL,0,1,2,3,4
HUMIDITY,5,6,7,8,9
WIND,10,11,12,13,14


In [ ]:
dframe1.rename(index=str.title, columns=str.upper)

,BERGEN,OSLO,TRONDHEIM,STAVANGER,KRISTIANSAND
Rainfall,0,1,2,3,4
Humidity,5,6,7,8,9
Wind,10,11,12,13,14


## Discretization and binning
Often when working with continuous datasets, we need to convert them into discrete or interval forms. Each interval is referred to as a bin, and hence the name binning comes into play:

In [1]:
height = [120, 122, 125, 127, 121, 123, 137, 131, 161, 145, 141, 132]

And we want to convert that dataset into intervals of 118 to 125, 126 to 135, 136 to 160, and finally 160 and higher.

To convert the preceding dataset into intervals, we can use the cut() method provided by the pandas library:

In [3]:
import pandas as pd
bins = [118, 125, 135, 160, 200]
category = pd.cut(height, bins)
category

[(118, 125], (118, 125], (118, 125], (125, 135], (118, 125], ..., (125, 135], (160, 200], (135, 160], (135, 160], (125, 135]]
Length: 12
Categories (4, interval[int64, right]): [(118, 125] < (125, 135] < (135, 160] < (160, 200]]

In [4]:
category2 = pd.cut(height, [118, 126, 136, 161, 200], right=False)
category2

[[118, 126), [118, 126), [118, 126), [126, 136), [118, 126), ..., [126, 136), [161, 200), [136, 161), [136, 161), [126, 136)]
Length: 12
Categories (4, interval[int64, left]): [[118, 126) < [126, 136) < [136, 161) < [161, 200)]

In [5]:
pd.value_counts(category)

/tmp/ipykernel_7873/2654243906.py:1: FutureWarning: pandas.value_counts is deprecated and will be removed in a future version. Use pd.Series(obj).value_counts() instead.
  pd.value_counts(category)


,count
"(118, 125]",5
"(125, 135]",3
"(135, 160]",3
"(160, 200]",1


In [6]:
bin_names = ['Short Height', 'Average height', 'Good Height', 'Taller']
pd.cut(height, bins, labels=bin_names)

['Short Height', 'Short Height', 'Short Height', 'Average height', 'Short Height', ..., 'Average height', 'Taller', 'Good Height', 'Good Height', 'Average height']
Length: 12
Categories (4, object): ['Short Height' < 'Average height' < 'Good Height' < 'Taller']

In [7]:
import numpy as np
pd.cut(np.random.rand(40), 5, precision=2)

[(0.2, 0.4], (0.0036, 0.2], (0.4, 0.59], (0.0036, 0.2], (0.59, 0.79], ..., (0.4, 0.59], (0.2, 0.4], (0.79, 0.98], (0.59, 0.79], (0.4, 0.59]]
Length: 40
Categories (5, interval[float64, right]): [(0.0036, 0.2] < (0.2, 0.4] < (0.4, 0.59] < (0.59, 0.79] <
                                           (0.79, 0.98]]

In [8]:
randomNumbers = np.random.rand(2000)
category3 = pd.qcut(randomNumbers, 4) # cut into quartiles
category3

[(-0.00051, 0.259], (0.259, 0.499], (0.259, 0.499], (0.755, 0.999], (0.755, 0.999], ..., (0.259, 0.499], (0.259, 0.499], (0.499, 0.755], (-0.00051, 0.259], (0.499, 0.755]]
Length: 2000
Categories (4, interval[float64, right]): [(-0.00051, 0.259] < (0.259, 0.499] < (0.499, 0.755] <
                                           (0.755, 0.999]]

In [9]:
pd.value_counts(category3)

/tmp/ipykernel_7873/647498483.py:1: FutureWarning: pandas.value_counts is deprecated and will be removed in a future version. Use pd.Series(obj).value_counts() instead.
  pd.value_counts(category3)


,count
"(-0.00051, 0.259]",500
"(0.259, 0.499]",500
"(0.499, 0.755]",500
"(0.755, 0.999]",500


# Outlier detection and filtering

In [10]:
df = pd.read_csv('https://raw.githubusercontent.com/PacktPublishing/hands-on-exploratory-data-analysis-with-python/master/Chapter%204/sales.csv')
df.head(10)

,Account,Company,Order,SKU,Country,Year,Quantity,UnitPrice,transactionComplete
0,123456779,Kulas Inc,99985,s9-supercomputer,Aruba,1981,5148,545,False
1,123456784,GitHub,99986,s4-supercomputer,Brazil,2001,3262,383,False
2,123456782,Kulas Inc,99990,s10-supercomputer,Montserrat,1973,9119,407,True
3,123456783,My SQ Man,99999,s1-supercomputer,El Salvador,2015,3097,615,False
4,123456787,ABC Dogma,99996,s6-supercomputer,Poland,1970,3356,91,True
5,123456778,Super Sexy Dingo,99996,s9-supercomputer,Costa Rica,2004,2474,136,True
6,123456783,ABC Dogma,99981,s11-supercomputer,Spain,2006,4081,195,False
7,123456785,ABC Dogma,99998,s9-supercomputer,Belarus,2015,6576,603,False
8,123456778,Loolo INC,99997,s8-supercomputer,Mauritius,1999,2460,36,False
9,123456775,Kulas Inc,99997,s7-supercomputer,French Guiana,2004,1831,664,True


In [11]:
df['TotalPrice'] = df['UnitPrice'] * df['Quantity']
df

,Account,Company,Order,SKU,Country,Year,Quantity,UnitPrice,transactionComplete,TotalPrice
0,123456779,Kulas Inc,99985,s9-supercomputer,Aruba,1981,5148,545,False,2805660
1,123456784,GitHub,99986,s4-supercomputer,Brazil,2001,3262,383,False,1249346
2,123456782,Kulas Inc,99990,s10-supercomputer,Montserrat,1973,9119,407,True,3711433
3,123456783,My SQ Man,99999,s1-supercomputer,El Salvador,2015,3097,615,False,1904655
4,123456787,ABC Dogma,99996,s6-supercomputer,Poland,1970,3356,91,True,305396
...,...,...,...,...,...,...,...,...,...,...
9995,123456784,Pryianka Ji,99987,s1-supercomputer,Jamaica,1983,886,475,False,420850
9996,123456775,Will LLC,99985,s3-supercomputer,Vietnam,2002,9995,302,True,3018490
9997,123456774,Kulas Inc,99982,s2-supercomputer,Northern Mariana Islands,1979,1421,249,True,353829
9998,123456781,Loolo INC,99986,s5-supercomputer,Mali,1991,2342,506,False,1185052


In [12]:
TotalTransaction = df["TotalPrice"]
TotalTransaction[np.abs(TotalTransaction) > 3000000]

,TotalPrice
2,3711433
7,3965328
13,4758900
15,5189372
17,3989325
...,...
9977,3475824
9984,5251134
9987,5670420
9991,5735513


In [13]:
df[np.abs(TotalTransaction) > 6741112]

,Account,Company,Order,SKU,Country,Year,Quantity,UnitPrice,transactionComplete,TotalPrice
818,123456781,Gen Power,99991,s1-supercomputer,Burkina Faso,1985,9693,696,False,6746328
1402,123456778,Will LLC,99985,s11-supercomputer,Austria,1990,9844,695,True,6841580
2242,123456770,Name IT,99997,s9-supercomputer,Myanmar,1979,9804,692,False,6784368
2876,123456772,Gen Power,99992,s10-supercomputer,Mali,2007,9935,679,False,6745865
3210,123456782,Loolo INC,99991,s8-supercomputer,Kuwait,2006,9886,692,False,6841112
3629,123456779,My SQ Man,99980,s3-supercomputer,Hong Kong,1994,9694,700,False,6785800
7674,123456781,Loolo INC,99989,s6-supercomputer,Sri Lanka,1994,9882,691,False,6828462
8645,123456789,Gen Power,99996,s11-supercomputer,Suriname,2005,9742,699,False,6809658
8684,123456785,Gen Power,99989,s2-supercomputer,Kenya,2013,9805,694,False,6804670


# Permutation and random sampling

In [14]:
dat = np.arange(80).reshape(10,8)
df = pd.DataFrame(dat)
df

,0,1,2,3,4,5,6,7
0,0,1,2,3,4,5,6,7
1,8,9,10,11,12,13,14,15
2,16,17,18,19,20,21,22,23
3,24,25,26,27,28,29,30,31
4,32,33,34,35,36,37,38,39
5,40,41,42,43,44,45,46,47
6,48,49,50,51,52,53,54,55
7,56,57,58,59,60,61,62,63
8,64,65,66,67,68,69,70,71
9,72,73,74,75,76,77,78,79


In [15]:
sampler = np.random.permutation(10)
sampler

array([9, 6, 2, 8, 3, 0, 7, 4, 1, 5])

In [16]:
df.take(sampler)

,0,1,2,3,4,5,6,7
9,72,73,74,75,76,77,78,79
6,48,49,50,51,52,53,54,55
2,16,17,18,19,20,21,22,23
8,64,65,66,67,68,69,70,71
3,24,25,26,27,28,29,30,31
0,0,1,2,3,4,5,6,7
7,56,57,58,59,60,61,62,63
4,32,33,34,35,36,37,38,39
1,8,9,10,11,12,13,14,15
5,40,41,42,43,44,45,46,47
